### ISSUE TITLE
`import/extensions` does not like patterns with `.` in them

### ISSUE THREAD

**Issue by osama-akhter** [import-js/eslint-plugin-import#3109]
It looks like the `import/extensions` rule only works with the last part of a file's extension. This is a problem if, for example, you want to allow importing `.d.ts` files. So, given a config like this

```json
"import/extensions": [
  "warn",
  "never",
  {
    "pattern": {
      "d.ts": "always",
      "json": "always"
    }
  }
]
```

This would allow `.json` extensions in imports but still not allow `.d.ts` (putting `.d.ts` also doesn't work). This can be important, for example, with [libraries that do not export all types](https://github.com/microsoft/TypeScript/issues/52363) in TypeScript. This requires you to do something like

```ts
declare module "foo" {
  export * from "foo/dist/types/index.d.ts";
}
```

Version Information
```json
"eslint": "^8.56.0",
"eslint-plugin-import": "^2.31.0",
```

**Comment by michaelfaith** [comment]
Similar root cause as https://github.com/import-js/eslint-plugin-import/issues/2147#issuecomment-1229007679
